allen-cell-animated/timelapse-colorizer
# feature: Accept binary `.parquet` data

Problem
=======

*Estimated review size: small, 10-15 minutes*

Closes #401, allowing the viewer to read binary data! See related PR in `colorizer-data` here that writes to the Parquet data format: https://github.com/allen-cell-animated/colorizer-data/pull/53

When testing on our exploratory dataset, the feature files went from 110 MB to only 10.8 MB using Parquet! This lets us open datasets MUCH faster and greatly improves the user experience.

Before: All data stored in JSON files, take up a large amount of space due to being saved as string characters.
```
{
  "data": [1.4, 2.3, 5.9,...] 
}
```

After: Data is stored in a columnar format in binary. Each value should take up no more than 32 bits, since we save values as 32-bit floats. Parquet does a lot of additional optimization so our storage size may be much smaller.
| data |
| --- |
| 1.4 |
| 2.3 | 
| 5.9 |
| ... |


Solution
========
- Adds the `hyparquet` and `hyparquet-compressors` packages to the project for Parquet file parsing.
- Updates the viewer to recognize `min` and `max` for feature data provided in the manifest, instead of needing it to be provided in the feature data file. (Viewer is still backwards-compatible with datasets that do, however.)
  - This lets us use parquet column datasets, which don't have a way to provide a min/max value.
- Refactors `JsonArrayLoader` into `UrlArrayLoader`, which loads and parses both `.json` and `.parquet` filetypes.

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Load the example here in your browser:  https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-409/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2Fparquet%2Fcollection.json&dataset=Medium+JSON
2. Switch between each dataset to observe the load times.

NOTE: I may swap this out for a link that accesses s3 data once I get access to our test bucket. For now, you can also try throttling the network activity from the network tab to see the difference.

NOTE: There's a known bug I've been trying to track down where Parquet datasets that load very quickly (like in this example) don't show the progress bar. I'm fairly certain this is because too many state updates are being requested, so React is blocking or discarding the state updates. It then only passes the updated progress prop when it reaches 100 and stops updating. There are potentially some hacky workarounds, like detecting the time of last update and then batching progress updates together until a certain time interval is hit, but I don't want to block this PR while I work on that. I also expect that this won't be an issue in environments where the dataset is loading more slowly (e.g. our s3 bucket).

Keyfiles (delete if not relevant):
-----------------------
1. main file/entry point
4. other important file

Thanks for contributing!
